src/Microsoft.AspNet.TestHost/WebSocketClient.cs

```diff
@@ -0,0 +1,177 @@
+            {
+                _websocketTcs = new TaskCompletionSource<System.Net.WebSockets.WebSocket>();
+
+                // HttpContext
+                FeatureCollection = new FeatureCollection();
+                var httpContext = new DefaultHttpContext(FeatureCollection);
+
+                // Request
+                httpContext.SetFeature<IHttpRequestFeature>(new RequestFeature());
+                var serverRequest = httpContext.Request;
+                serverRequest.Protocol = "HTTP/" + clientRequest.Version.ToString(2);
+                serverRequest.Scheme = clientRequest.RequestUri.Scheme;
+                serverRequest.Method = clientRequest.Method.ToString();
+                var fullPath = PathString.FromUriComponent(clientRequest.RequestUri);
+                PathString remainder;
+                if (fullPath.StartsWithSegments(pathBase, out remainder))
+                {
+                    serverRequest.PathBase = pathBase;
+                    serverRequest.Path = remainder;
+                }
+                else
+                {
+                    serverRequest.PathBase = PathString.Empty;
+                    serverRequest.Path = fullPath;
+                }
+                serverRequest.QueryString = QueryString.FromUriComponent(clientRequest.RequestUri);
+                foreach (var header in clientRequest.Headers)
+                {
+                    serverRequest.Headers.AppendValues(header.Key, header.Value.ToArray());
+                }
+                var requestContent = clientRequest.Content;
+                if (requestContent != null)
+                {
+                    foreach (var header in clientRequest.Content.Headers)
+                    {
+                        serverRequest.Headers.AppendValues(header.Key, header.Value.ToArray());
+                    }
+                }
+                serverRequest.Body = Stream.Null;
+
+                // Response
+                httpContext.SetFeature<IHttpResponseFeature>(new ResponseFeature());
+                httpContext.Response.Body = Stream.Null;
+                httpContext.Response.StatusCode = 200;
+
+                // WebSocket
+                httpContext.SetFeature<IHttpWebSocketFeature>(this);
+            }
+
+            public void PipelineComplete()
+            {
+                PipelineFailed(null);
+            }
+
+            public void PipelineFailed(Exception ex)
+            {
+                _websocketTcs.TrySetException(new InvalidOperationException("The websocket was not accepted.", ex));
+            }
+
+            bool IHttpWebSocketFeature.IsWebSocketRequest
+            {
+                get
+                {
+                    return true;
+                }
+            }
+
+            Task<System.Net.WebSockets.WebSocket> IHttpWebSocketFeature.AcceptAsync(WebSocketAcceptContext context)
+            {
+                var websockets = WebSocket.CreatePair(context.SubProtocol);
```

Comment:
Store both of these and abort/dispose them when _next finishes or throws (e.g. PipelineComplete or PipelineFailed).